File: src/main/java/rars/venus/editors/TextEditingArea.java
package rars.venus.editors;

import kotlin.Pair;
import org.jetbrains.annotations.NotNull;
import rars.riscv.lang.lexing.RVTokenType;

import javax.swing.text.Caret;
import javax.swing.text.Document;
import java.awt.*;

/**
 * Specifies capabilities that any test editor used in MARS must have.
 */
public interface TextEditingArea {

    void copy();

    void cut();

    @NotNull FindReplaceResult doFindText(String find, boolean caseSensitive);

    @NotNull FindReplaceResult doReplace(String find, String replace, boolean caseSensitive);

    int doReplaceAll(String find, String replace, boolean caseSensitive);

    Document getDocument();

    void select(int selectionStart, int selectionEnd);

    void selectLine(int lineNumber);

    String getText();

    void setText(String text);

    void paste();

    void setEditable(boolean editable);

    Font getFont();

    void setFont(Font f);

    void requestFocusInWindow();

    void setForeground(Color c);

    void setBackground(Color c);

    void setSelectionColor(Color c);

    void setCaretColor(Color c);

    void setLineHighlightColor(Color c);

    @NotNull Caret getCaret();

    void setEnabled(boolean enabled);

    void redo();

    void setSourceCode(String code, boolean editable);

    void undo();

    void discardAllUndoableEdits();

    void setLineHighlightEnabled(boolean highlight);

    int getCaretBlinkRate();

    void setCaretBlinkRate(int rate);

    int getTabSize();

    void setTabSize(int chars);

    @NotNull Pair<@NotNull Integer, @NotNull Integer> getCaretPosition();

    Component getOuterComponent();

    boolean canUndo();

    boolean canRedo();

    @NotNull EditorTheme getTheme();

    void setTheme(final @NotNull EditorTheme theme);

    void setTokenStyle(final @NotNull RVTokenType type, final @NotNull TokenStyle style);

    // Used by Find/Replace
    enum FindReplaceResult {
        TEXT_NOT_FOUND,
        TEXT_FOUND,
        TEXT_REPLACED_FOUND_NEXT,
        TEXT_REPLACED_NOT_FOUND_NEXT
    }
}

File: src/main/java/rars/venus/editors/TextEditingArea.kt
package rars.venus.editors

import rars.riscv.lang.lexing.RVTokenType
import rars.settings.AllSettings
import rars.venus.editors.rsyntaxtextarea.RSyntaxTextAreaBasedEditor
import java.awt.Color
import java.awt.Component
import java.awt.Font
import javax.swing.text.Caret
import javax.swing.text.Document

/**
 * Specifies capabilities that any test editor used in MARS must have.
 */
interface TextEditingArea {
    val document: Document

    var text: String

    var isEditable: Boolean

    var font: Font

    var theme: EditorTheme

    var foreground: Color

    var background: Color

    var selectionColor: Color

    var caretColor: Color

    var lineHighlightColor: Color

    val caret: Caret

    var isEnabled: Boolean

    fun setSourceCode(code: String, editable: Boolean)

    var lineHighlightEnabled: Boolean

    var caretBlinkRate: Int

    var tabSize: Int

    val caretPosition: Pair<Int, Int>

    val outerComponent: Component

    fun copy()

    fun cut()

    fun paste()

    fun undo()

    fun redo()

    fun canUndo(): Boolean

    fun canRedo(): Boolean

    fun discardAllUndoableEdits()

    fun doFindText(find: String, caseSensitive: Boolean): FindReplaceResult

    fun doReplace(find: String, replace: String, caseSensitive: Boolean): FindReplaceResult

    fun doReplaceAll(find: String, replace: String, caseSensitive: Boolean): Int

    fun select(selectionStart: Int, selectionEnd: Int)

    fun selectLine(lineNumber: Int)

    fun requestFocusInWindow()


    fun setTokenStyle(type: RVTokenType, style: TokenStyle)

    // Used by Find/Replace
    enum class FindReplaceResult {
        TEXT_NOT_FOUND,
        TEXT_FOUND,
        TEXT_REPLACED_FOUND_NEXT,
        TEXT_REPLACED_NOT_FOUND_NEXT
    }
}

fun createTextEditingArea(allSettings: AllSettings): TextEditingArea = RSyntaxTextAreaBasedEditor(
    allSettings.editorThemeSettings.currentTheme.toEditorTheme(),
    allSettings.fontSettings
)